overlord/devicestate/devicestate_cloudinit_test.go

```diff
@@ -0,0 +1,951 @@
+
+	r = devicestate.MockRestrictCloudInit(func(sysconfig.CloudInitState, *sysconfig.CloudInitRestrictOptions) (sysconfig.CloudInitRestrictionResult, error) {
+		c.Error("EnsureCloudInitRestricted should not have restricted cloud-init when not seeded")
+		return sysconfig.CloudInitRestrictionResult{}, fmt.Errorf("broken")
+	})
+	defer r()
+
+	err := devicestate.EnsureCloudInitRestricted(s.mgr)
+	c.Assert(err, IsNil)
+}
+
+func (s *cloudInitSuite) TestCloudInitAlreadyEnsuredRestrictedDoesNothing(c *C) {
+	n := 0
+
+	// mock that it was restricted so that we set the internal bool to say it
+	// already ran
+	r := devicestate.MockCloudInitStatus(func() (sysconfig.CloudInitState, error) {
+		n++
+		switch n {
+		case 1:
+			return sysconfig.CloudInitRestrictedBySnapd, nil
+		default:
+			c.Error("EnsureCloudInitRestricted should not have checked cloud-init status again")
+			return sysconfig.CloudInitRestrictedBySnapd, fmt.Errorf("test broken")
+		}
+	})
+	defer r()
+
+	// run it once to set the internal bool
+	err := devicestate.EnsureCloudInitRestricted(s.mgr)
+	c.Assert(err, IsNil)
+
+	c.Assert(n, Equals, 1)
+
+	// it should run again without checking anything
+	err = devicestate.EnsureCloudInitRestricted(s.mgr)
+	c.Assert(err, IsNil)
+
+	c.Assert(n, Equals, 1)
+}
+
+func (s *cloudInitSuite) TestCloudInitDeviceManagerEnsureRestrictsCloudInit(c *C) {
+	st := s.o.State()
+	st.Lock()
+	// avoid device registration
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Serial: "123",
+	})
+	st.Unlock()
+	n := 0
+
+	// mock that it was restricted so that we set the internal bool to say it
+	// already ran
+	r := devicestate.MockCloudInitStatus(func() (sysconfig.CloudInitState, error) {
+		n++
+		switch n {
+		case 1:
+			return sysconfig.CloudInitRestrictedBySnapd, nil
+		default:
+			c.Error("EnsureCloudInitRestricted should not have checked cloud-init status again")
+			return sysconfig.CloudInitRestrictedBySnapd, fmt.Errorf("test broken")
+		}
+	})
+	defer r()
+
+	// run it once to set the internal bool
+	err := s.mgr.Ensure()
+	c.Assert(err, IsNil)
+	c.Assert(n, Equals, 1)
+
+	// running again is still okay and won't call CloudInitStatus again
+	err = s.mgr.Ensure()
+	c.Assert(err, IsNil)
+	c.Assert(n, Equals, 1)
+}
+
+func (s *cloudInitSuite) TestCloudInitAlreadyRestrictedDoesNothing(c *C) {
+	r := devicestate.MockCloudInitStatus(func() (sysconfig.CloudInitState, error) {
+		return sysconfig.CloudInitRestrictedBySnapd, nil
+	})
+	defer r()
+
+	r = devicestate.MockRestrictCloudInit(func(sysconfig.CloudInitState, *sysconfig.CloudInitRestrictOptions) (sysconfig.CloudInitRestrictionResult, error) {
+		c.Error("EnsureCloudInitRestricted should not have restricted cloud-init when already restricted")
+		return sysconfig.CloudInitRestrictionResult{}, fmt.Errorf("broken")
+	})
+	defer r()
+
+	err := devicestate.EnsureCloudInitRestricted(s.mgr)
+	c.Assert(err, IsNil)
+}
+
+func (s *cloudInitSuite) TestCloudInitAlreadyDisabledDoesNothing(c *C) {
```

Comment:
it occurs to me a variant of this test with the zzzz_snapd.cfg file existing would be nice to have as well